src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -221,37 +216,66 @@ ClientImpl::ClientImpl(ClientConfig config)
 }
 
 
-std::string ClientImpl::make_user_agent_string(ClientConfig& config)
+void ClientImpl::post(SyncSocketProvider::FunctionHandler&& handler)
 {
-    std::string platform_info = std::move(config.user_agent_platform_info);
-    if (platform_info.empty())
-        platform_info = util::get_platform_info(); // Throws
-    std::ostringstream out;
-    out << "RealmSync/" REALM_VERSION_STRING " (" << platform_info << ")"; // Throws
-    if (!config.user_agent_application_info.empty())
-        out << " " << config.user_agent_application_info; // Throws
-    return out.str();                                     // Throws
+    REALM_ASSERT(m_socket_provider);
+    {
+        std::lock_guard lock(m_drain_mutex);
+        ++m_outstanding_posts;
+        m_drained = false;
+    }
+    m_socket_provider->post([handler = std::move(handler), this](Status status) {
+        handler(status);
+
+        std::lock_guard lock(m_drain_mutex);
+        --m_outstanding_posts;
+        m_drain_cv.notify_all();
+    });
 }
 
 
-void ClientImpl::post(SyncSocketProvider::FunctionHandler&& handler)
+void ClientImpl::drain_connections()
 {
-    REALM_ASSERT(m_socket_provider);
-    m_socket_provider->post(std::move(handler));
+    logger.debug("Draining connections during sync client shutdown");
+    for (auto& server_slot_pair : m_server_slots) {
+        auto& server_slot = server_slot_pair.second;
+
+        if (server_slot.connection) {
+            auto& conn = server_slot.connection;
+            conn->force_close();
+        }
+        else {
+            for (auto& conn_pair : server_slot.alt_connections) {
+                conn_pair.second->force_close();
+            }
+        }
+    }
 }
 
 
 SyncSocketProvider::SyncTimer ClientImpl::create_timer(std::chrono::milliseconds delay,
                                                        SyncSocketProvider::FunctionHandler&& handler)
 {
     REALM_ASSERT(m_socket_provider);
-    return m_socket_provider->create_timer(delay, std::move(handler));
+    {
+        std::lock_guard lock(m_drain_mutex);
+        ++m_outstanding_posts;
+        m_drained = false;
+    }
+    return m_socket_provider->create_timer(delay, [handler = std::move(handler), this](Status status) {
```

Comment:
Should the create timer operation be guaranteed to be performed on the event loop?
The `DefaultSocketProvider::Timer` class could be updated to post the `async_wait` and `cancel` operations on the event loop...

Reply:
I don't think the changes in this PR really change the behavior of timers overall, and I'd rather not introduce that uncertainty here.